### PR TITLE
fix: --pod-max-pids upgrade to 1.14 scenarios

### LINKED ISSUE #1270: Pods crash with: pthread_create() failed (11: Resource temporarily unavailable) after cluster upgrade
**Is this a request for help?**: Yes

---

**Is this an ISSUE or FEATURE REQUEST?** (choose one): Issue

---

**What version of aks-engine?**: 0.35.1

---

**Kubernetes version**: 1.14.1

**What happened**:
After upgrading our QA cluster using AKS-Engine 0.35.1 from K8s 1.11.6 via 1.12.8, 1.13.5 to 1.14.1 workload pods do not start anymore or crash after a while showing the error "pthread_create() failed (11: Resource temporarily unavailable)" or similar.
Pods crashing are for example RabbitMQ or Nginx-Ingress controller.

kubectl describe pod shows:

> Name:               rabbitmq-rabbitmq-ha-0
> Namespace:          qa
> Priority:           0
> PriorityClassName:  <none>
> Node:               k8s-static-11480702-vmss000013/10.239.0.18
> Start Time:         Thu, 09 May 2019 16:55:53 +0200
> Labels:             component=rabbitmq
>                     controller-revision-hash=rabbitmq-rabbitmq-ha-5cc8495b8f
>                     statefulset.kubernetes.io/pod-name=rabbitmq-rabbitmq-ha-0
>                     type=server
> Annotations:        cni.projectcalico.org/podIP=10.244.92.4/32
> Status:             Running
> IP:                 10.244.92.4
> Controlled By:      StatefulSet/rabbitmq-rabbitmq-ha
> Init Containers:
>   copy-rabbitmq-config:
>     Container ID:  docker://5816dd3e044a0ddc497dc0de1cb3736020f9153ea1f38752e022a22ceb014877
>     Image:         qnowsacr.azurecr.io/external/busybox:1.29.2@sha256:3058e3a1129c64da64d5c7889e6eedb0> 666262d7ee69b289f2d4379f69362383
>     Image ID:      docker-pullable://qnowsacr.azurecr.io/external/busybox@sha256:3058e3a1129c64da64d5c7889e6eedb0666262d7ee69b289f2d4379f69362383
>     Port:          <none>
>     Host Port:     <none>
>     Command:
>       sh
>       -c
>       cp /configmap/* /etc/rabbitmq; rm -f /var/lib/rabbitmq/.erlang.cookie
>     State:          Terminated
>      Reason:       Completed
>      Exit Code:    0
>      Started:      Thu, 09 May 2019 16:56:34 +0200
>      Finished:     Thu, 09 May 2019 16:56:34 +0200
>    Ready:          True
>    Restart Count:  0
>    Environment:    <none>
>    Mounts:
>      /configmap from configmap (rw)
>      /etc/rabbitmq from config (rw)
>      /var/lib/rabbitmq from data (rw)
>      /var/run/secrets/kubernetes.io/serviceaccount from rabbitmq-token-6qx8h (ro)
> Containers:
>   rabbitmq-ha:
 >    Container ID:   docker://3313cb9f1cbdbe67c5fb19bfb3c2ade0eaf95456917e9aa04636c1c5740b009b
 >    Image:          qnowsacr.azurecr.io/external/rabbitmq:3.7.8-management-alpine@sha256:062935e77e35e8e7d677decf841cabf0c7c84d80d3d8ea362ad612c2d3c05e70
>     Image ID:       docker-pullable://qnowsacr.azurecr.io/external/rabbitmq@sha256:062935e77e35e8e7d677decf841cabf0c7c84d80d3d8ea362ad612c2d3c05e70
>     Ports:          4369/TCP, 5672/TCP, 15672/TCP
>     Host Ports:     0/TCP, 0/TCP, 0/TCP
>     State:          Running
>       Started:      Thu, 09 May 2019 16:56:42 +0200
>     Ready:          True
>     Restart Count:  0
>     Limits:
>       cpu:     8
>       memory:  55Gi
>     Requests:
>       cpu:     8
>       memory:  55Gi
>     Environment:
>       MY_POD_NAME:             rabbitmq-rabbitmq-ha-0 (v1:metadata.name)
>       RABBITMQ_USE_LONGNAME:   true
>       RABBITMQ_NODENAME:       rabbit@$(MY_POD_NAME).rabbitmq-rabbitmq-ha-discovery.qa.svc.cluster.local
>       K8S_HOSTNAME_SUFFIX:     .rabbitmq-rabbitmq-ha-discovery.qa.svc.cluster.local
>       K8S_SERVICE_NAME:        rabbitmq-rabbitmq-ha-discovery
>       RABBITMQ_ERLANG_COOKIE:  <set to the key 'rabbitmq-erlang-cookie' in secret 'rabbitmq-provided'>   Optional: false
>       RABBITMQ_DEFAULT_USER:   <set to the key 'rabbitmq-admin-username' in secret 'rabbitmq-provided'>  Optional: false
>      RABBITMQ_DEFAULT_PASS:   <set to the key 'rabbitmq-admin-password' in secret 'rabbitmq-provided'>  Optional: false
>       RABBITMQ_DEFAULT_VHOST:  /
>     Mounts:
>       /etc/rabbitmq from config (rw)
>       /var/lib/rabbitmq from data (rw)
>       /var/run/secrets/kubernetes.io/serviceaccount from rabbitmq-token-6qx8h (ro)
>   rabbitmq-ha-exporter:
>     Container ID:   docker://ffe6e5f755d9b77eb62addbccfe557756c21c73debb4274915e193f72fcbc4f6
>     Image:          qnowsacr.azurecr.io/external/rabbitmq-exporter:v0.29.0@sha256:424c036132bfe7f31674eb9a4d0c60395ec6fd794ab08e5eda6f206e13984b21
>     Image ID:       docker-pullable://qnowsacr.azurecr.io/external/rabbitmq-exporter@sha256:424c036132bfe7f31674eb9a4d0c60395ec6fd794ab08e5eda6f206e13984b21
>     Port:           9419/TCP
>     Host Port:      0/TCP
>     State:          Waiting
>       Reason:       CrashLoopBackOff
>     Last State:     Terminated
>       Reason:       ContainerCannotRun
>       Message:      OCI runtime create failed: container_linux.go:344: starting container process caused "process_linux.go:424: container init caused \"read init-p: connection reset by peer\"": unknown
>       Exit Code:    128
>       Started:      Thu, 09 May 2019 16:56:58 +0200
>       Finished:     Thu, 09 May 2019 16:56:58 +0200
>     Ready:          False
>     Restart Count:  2
>     Environment:
>       PUBLISH_PORT:         9419
>       RABBIT_CAPABILITIES:  bert,no_sort
>       RABBIT_USER:          admin
>       RABBIT_PASSWORD:      <set to the key 'rabbitmq-admin-password' in secret 'rabbitmq-provided'>  Optional: false
>     Mounts:
>       /var/run/secrets/kubernetes.io/serviceaccount from rabbitmq-token-6qx8h (ro)
> Conditions:
>   Type              Status
>   Initialized       True 
>   Ready             False 
>   ContainersReady   False 
>   PodScheduled      True 
> Volumes:
>   data:
>     Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
>     ClaimName:  data-rabbitmq-rabbitmq-ha-0
>     ReadOnly:   false
>   config:
>     Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
>     Medium:  
>   configmap:
>     Type:      ConfigMap (a volume populated by a ConfigMap)
>     Name:      rabbitmq
>     Optional:  false
>   rabbitmq-token-6qx8h:
>     Type:        Secret (a volume populated by a Secret)
>     SecretName:  rabbitmq-token-6qx8h
>     Optional:    false
> QoS Class:       Burstable
> Node-Selectors:  agentpool=static
> Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
>                  node.kubernetes.io/unreachable:NoExecute for 300s
> Events:
>   Type     Reason                  Age                From                                     Message
>   ----     ------                  ----               ----                                     -------
>   Normal   Scheduled               1m                 default-scheduler                        Successfully assigned qa/rabbitmq-rabbitmq-ha-0 to k8s-static-11480702-vmss000013
>   Normal   SuccessfulAttachVolume  1m                 attachdetach-controller                  AttachVolume.Attach succeeded for volume "pvc-3f916d40-5615-11e9-9006-000d3ab8e732"
>   Normal   Pulling                 52s                kubelet, k8s-static-11480702-vmss000013  Pulling image "qnowsacr.azurecr.io/external/busybox:1.29.2@sha256:3058e3a1129c64da64d5c7889e6eedb0666262d7ee69b289f2d4379f69362383"
>   Normal   Pulled                  51s                kubelet, k8s-static-11480702-vmss000013  Successfully pulled image "qnowsacr.azurecr.io/external/busybox:1.29.2@sha256:3058e3a1129c64da64d5c7889e6eedb0666262d7ee69b289f2d4379f69362383"
>   Normal   Created                 51s                kubelet, k8s-static-11480702-vmss000013  Created container copy-rabbitmq-config
>   Normal   Started                 51s                kubelet, k8s-static-11480702-vmss000013  Started container copy-rabbitmq-config
>   Normal   Pulling                 50s                kubelet, k8s-static-11480702-vmss000013  Pulling image "qnowsacr.azurecr.io/external/rabbitmq:3.7.8-management-alpine@sha256:062935e77e35e8e7d677decf841cabf0c7c84d80d3d8ea362ad612c2d3c05e70"
>   Normal   Pulled                  45s                kubelet, k8s-static-11480702-vmss000013  Successfully pulled image "qnowsacr.azurecr.io/external/rabbitmq:3.7.8-management-alpine@sha256:062935e77e35e8e7d677decf841cabf0c7c84d80d3d8ea362ad612c2d3c05e70"
>   Normal   Created                 44s                kubelet, k8s-static-11480702-vmss000013  Created container rabbitmq-ha
>   Normal   Started                 43s                kubelet, k8s-static-11480702-vmss000013  Started container rabbitmq-ha
>   Normal   Pulling                 27s (x3 over 43s)  kubelet, k8s-static-11480702-vmss000013  Pulling image "qnowsacr.azurecr.io/external/rabbitmq-exporter:v0.29.0@sha256:424c036132bfe7f31674eb9a4d0c60395ec6fd794ab08e5eda6f206e13984b21"
>   Normal   Pulled                  27s (x3 over 42s)  kubelet, k8s-static-11480702-vmss000013  Successfully pulled image "qnowsacr.azurecr.io/external/rabbitmq-exporter:v0.29.0@sha256:424c036132bfe7f31674eb9a4d0c60395ec6fd794ab08e5eda6f206e13984b21"
>   Normal   Created                 27s (x3 over 42s)  kubelet, k8s-static-11480702-vmss000013  Created container rabbitmq-ha-exporter
>   Warning  Failed                  26s (x3 over 41s)  kubelet, k8s-static-11480702-vmss000013  Error: failed to start container "rabbitmq-ha-exporter": Error response from daemon: OCI runtime create failed: container_linux.go:344: starting container process caused "process_linux.go:424: container init caused \"read init-p: connection reset by peer\"": unknown
>   Warning  BackOff                 8s (x3 over 12s)   kubelet, k8s-static-11480702-vmss000013  Back-off restarting failed container

Most of the systems pods run, but some of them (calico for example) do crash too.

Cluster was initially set up with ACS-Engine 0.24.1 (k8s 1.10.9) and upgraded successfully to k8s 1.11.6 with AKS-Engine 0.29.1.

**What you expected to happen**:

Cluster running normally, with our workloads, that used to be running fine until upgrading with 035.1.

**How to reproduce it** (as minimally and precisely as possible):
Initial setup of Cluster with acs-engine 0.24.1:
- Kubernetes 1.10.9
- Private cluster
- RBAC enabled
- Calico Networkpolicy
- AAD Integration enabled
- Disable addons:
  - blobfuse-flexvolume
  - smb-flexvolume
- 3 Master
- 3 Agentpools as vmss

Upgrade to 1.11.5 with AKS-Engine 0.29.1 (successful)
Upgrade to 1.11.6 with AKS-Engine 0.29.1 (successful)
Upgrade to 1.14.1 via 1.12.8 and 1.13.5 (three steps) with AKS-Engine 0.35.1 

**Anything else we need to know**:
Luckily this was our test to upgrade on our staging environment before doing the actual upgrade of our PROD envionment.

### PR DESCRIPTION
**Pull request by jackfrancis**
**Reason for Change**:

When upgrading to 1.14 from older AKS Engine versions, the api model will declare a `--pod-max-pids` kubeletConfig value of 100. Because the required feature flag to actually enforce that setting was disabled by default *until* 1.14, in practice this means that upgraded clusters will suddenly have a reduced PIDs/pods operational configuration, which will likely break many workload scenarios.

**Issue Fixed**:

Fixes #1270

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: pkg/api/defaults-kubelet.go [hunk 1]
```diff
@@ -133,6 +133,18 @@ func (cs *ContainerService) setKubeletConfig() {
...
 	}
 
+	if isUpgrade && common.IsKubernetesVersionGe(o.OrchestratorVersion, "1.14.0") {
+		hasSupportPodPidsLimitFeatureGate := strings.Contains(o.KubernetesConfig.KubeletConfig["--feature-gates"], "SupportPodPidsLimit=true")
+		podMaxPids, _ := strconv.Atoi(o.KubernetesConfig.KubeletConfig["--pod-max-pids"])
+		if podMaxPids > 0 {
+			// If we don't have an explicit SupportPodPidsLimit=true, disable --pod-max-pids by setting to -1
+			// To prevent older clusters from inheriting SupportPodPidsLimit=true implicitly starting w/ 1.14.0
+			if !hasSupportPodPidsLimitFeatureGate {
+				o.KubernetesConfig.KubeletConfig["--pod-max-pids"] = strconv.Itoa(-1)
```
Comment (CecileRobertMichon) [inline pkg/api/defaults-kubelet.go]: isn't the default value already -1? https://github.com/Azure/aks-engine/blob/master/pkg/api/defaults-kubelet.go#L78

Comment (CecileRobertMichon) [reply pkg/api/defaults-kubelet.go]: Oh I think I understand what's going on... we aren't overriding the default value during upgrades and this default was previously 100

Comment (jackfrancis) [reply pkg/api/defaults-kubelet.go]: Yeah, it's basically a timebomb. A config value of "100", that, if enabled, would demonstrably change the operational reality of cluster behaviors. Then it gets enabled (in 1.14.1).

This is why we did this:

https://github.com/Azure/aks-engine/pull/1508

We just didn't address the back-compat ramifications.

Comment (jackfrancis) [reply pkg/api/defaults-kubelet.go]: Whoops I meant https://github.com/Azure/aks-engine/pull/1126 changed the value to -1

**Pull request merged**